pkg/entrypoint/run.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package entrypoint
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strconv"
+	"syscall"
+
+	"go.uber.org/zap"
+)
+
+const (
+	// InternalErrorCode is what we write to the marker file to
+	// indicate that we failed to start the wrapped command
+	InternalErrorCode = 127
+
+	// DefaultShouldWaitForPrevStep is the default value for whether the
+	// command the entrypoint binary will launch should wait for a "finished"
+	// signal from another job.  This allows ordering steps
+	DefaultShouldWaitForPrevStep = false
+
+	// DefaultShouldRunPostRun is the default value for whether after the
+	// command finishes, it should send a "finished" signal that other waiting
+	// jobs might be relying on to begin. This allows ordering steps
+	DefaultShouldRunPostRun = false
+
+	// DefaultPreRunFile is the name of the file that a
+	// waiting job will be waiting to read before it runs.
+	DefaultPreRunFile = "0"
+
+	// DefaultPostRunFile is the name of the file that a
+	// finishing job will be write after it successfully completes.
+	DefaultPostRunFile = "1"
+)
+
+// Run executes the test process then writes the exit code to the marker file.
+// This function returns the status code that should be passed to os.Exit().
+func (o Options) Run(logger *zap.SugaredLogger) int {
+	defer logger.Sync()
+
+	code, err := o.ExecuteProcess()
+	if err != nil {
+		logger.Errorf("error executing user process: %v", err)
+	}
+	return code
+}
+
+// ExecuteProcess creates the artifact directory then executes the process as
+// configured, writing the output to the process log.
+func (o Options) ExecuteProcess() (int, error) {
+	var commandErr error
+
+	// wait for previous step if specified
+	if o.ShouldWaitForPrevStep {
+		done := make(chan error)
+		go func() {
+			done <- o.waitForPrevStep()
+		}()
+	}
+
+	var arguments []string
+	if len(o.Args) > 1 {
+		arguments = o.Args[1:]
+	}
+	command := exec.Command(o.Args[0], arguments...)
+	if err := command.Start(); err != nil {
+		return InternalErrorCode, fmt.Errorf("could not start the process: %v", err)
+	}
+
+	// execute the user specified command
+	done := make(chan error)
+	go func() {
+		done <- command.Wait()
+	}()
+	select {
+	case commandErr = <-done:
+	}
+
+	var returnCode int
+	if status, ok := command.ProcessState.Sys().(syscall.WaitStatus); ok {
+		returnCode = status.ExitStatus()
```

Comment:
> what does it mean if ok is true 

It verifies that output of `command.ProcessState.Sys()` type assertion to `syscall.WaitStatus` type succeeded.

It might be a good idea to declare some constants for `0` and `1`

Reply:
> commandErr is not nil? 

`Run()` function catches this error and logs in line 61.

Reply:
going to mark these resolved, doesn't seem like there is anything to do for this section. Feel free to comment if there is something to be done and I missed it.